osu.Game/Rulesets/Scoring/JudgementProcessor.cs

```diff
@@ -174,22 +162,52 @@ void simulate(HitObject obj)
             IsSimulating = false;
         }
 
-        protected override void Update()
+        /// <summary>
+        /// Enumerates all <see cref="HitObject"/>s in the given <paramref name="beatmap"/> in the order in which they are to be judged.
+        /// Used in <see cref="SimulateAutoplay"/>.
+        /// </summary>
+        /// <remarks>
+        /// In Score V2, the score awarded for each object includes a component based on the combo value after the judgement of that object.
+        /// This means that the score is dependent on the order of evaluation of judgements.
+        /// This method is provided so that rulesets can specify custom ordering that is correct for them and matches processing order during actual gameplay.
+        /// </remarks>
+        protected virtual IEnumerable<HitObject> EnumerateHitObjects(IBeatmap beatmap)
+            => enumerateRecursively(beatmap.HitObjects);
```

Comment:
Someone reported that a [graved osu! map](https://discord.com/channels/188630481301012481/1097318920991559880/1118985874001113140) also has an issue where 1M score is unachievable, and it's because this implementation is still naive (relies on no concurrent objects to work correctly). I could go and fix that case here by partially mirroring the mania implementation (flatten all objects, sort by end time) but honestly not sure it's worth the performance hit for a graved map.

Reply:
I wouldn't be against this since it's to guarantee correctness. The performance hit shouldn't be noticeable. But the current changeset is also fine for now.